presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java

```diff
@@ -158,12 +159,8 @@ public Pattern getPattern()
                 newSourceBuilder.build(),
                 inputsBuilder.build());
 
-        if (!result.getOutputSymbols().equals(project.getOutputSymbols())) {
-            // we need to strip unnecessary symbols (hash, partitioning columns).
-            result = new ProjectNode(idAllocator.getNextId(), result, identity(project.getOutputSymbols()));
-        }
-
-        return Optional.of(result);
+        // we need to strip unnecessary symbols (hash, partitioning columns).
+        return Optional.of(restrictOutputs(idAllocator, result, ImmutableSet.copyOf(project.getOutputSymbols())).orElse(result));
```

Comment:
Yes.  The outer Optional is about whether the PushProjectionThroughExchange fired, which at this point in the code should always be true, and the inner Optional is about whether a project-off was needed.